src/main/scala/temple/DSL/semantics/ServiceRenamer.scala

```diff
@@ -0,0 +1,88 @@
+package temple.DSL.semantics
+
+import temple.ast._
+
+case class ServiceRenamer(renaming: Map[String, String]) {
+
+  private def renameTargetBlock(block: TargetBlock): TargetBlock =
+    // currently `identity`, as language is the only metadata
+    TargetBlock(block.metadata.map {
+      case language: Metadata.TargetLanguage => language
+    })
+
+  private def renameProjectBlock(block: ProjectBlock): ProjectBlock =
+    // currently `identity`, as no metadata contains a service name
+    ProjectBlock(
+      block.metadata.map {
+        case language: Metadata.ServiceLanguage => language
+        case provider: Metadata.Provider        => provider
+        case database: Metadata.Database        => database
+        case readable: Metadata.Readable        => readable
+        case writable: Metadata.Writable        => writable
+      },
+    )
+
+  private def renameTargetBlocks(targets: Map[String, TargetBlock]): Map[String, TargetBlock] =
+    targets.view.mapValues(renameTargetBlock).toMap
+
+  private def renameServiceMetadata(metadata: Metadata.ServiceMetadata): Metadata.ServiceMetadata = metadata match {
+    // rename any services referenced in #uses
+    case Metadata.Uses(services) => Metadata.Uses(services.map(renaming))
```

Comment:
This isn't too clear what's going on? You're using a map lookup as a function? Then what happens if `renaming` doesn't contain that key?

Reply:
Yes, maps are functions. I could make it clearer if you’d like. If it doesn’t include the key it will throw a (maybe not particularly easy to read) error, I’ll update that to make it more readable.